Support perform_actions

aaronchongth opened the pull request:
## New feature implementation

### Implemented feature

Fixes https://github.com/open-rmf/free_fleet/issues/116

### Implementation description

Referencing
* https://github.com/open-rmf/fleet_adapter_template
* https://github.com/open-rmf/rmf_demos/tree/main/rmf_demos_fleet_adapter
* https://github.com/open-rmf/fleet_adapter_mir

This implements custom performable actions for the `free_fleet_adapter`.

* actions are implemented and loaded at runtime as plugins
* example actions `hello_world` and `delayed_hello_world`
* tests

### Testing

Start the `free_fleet_adapter` as per usual with the usual instructions.

Custom actions can be triggered using scripts from `rmf_demos_tasks`,

```bash
source ~/ff_ws/install/setup.bash
export ROS_DOMAIN_ID=55

# hello_world
ros2 run rmf_demos_tasks dispatch_action -st 0 -a hello_world

# hello_world with user name in description
ros2 run rmf_demos_tasks dispatch_action -st 0 -a hello_world \
  -ad '{"user": "John Doe"}'

# delayed_hello_world, default as 5 seconds
ros2 run rmf_demos_tasks dispatch_action -st 0 -a delayed_hello_world

# delayed_hello_world with user name and custom wait duration in description
ros2 run rmf_demos_tasks dispatch_action -st 0 -a delayed_hello_world \
  -ad '{"user": "Jane Doe", "wait_duration_sec": 20}'

# While a `delayed_hello_world` is ongoing, users can also trigger a cancellation manually on the action, which will cause the task to be cancelled as well.
ros2 topic pub --once  /cancel_delayed_hello_world std_msgs/msg/Empty "{}"
```

With these few examples so far, we show that
* custom actions can be implemented in a different python package, and imported in during runtime
* actions can be completed immediately, or run for an unspecified duration
* while the actions are ongoing, scenarios/situations/users can trigger a cancellation, or the action can be implemented to fail too

### Considerations

Technically, with nav2, we can implement a custom action for any behavior tree that is loaded by the robot, which includes the basic `navigate_to_pose` too. So even the navigation command can be wrapped up in an action.

Once the abstraction has been approved, we can implement some basic actions that trigger some of the out-of-the-box nav2 actions, e.g. localize, dock, spin around (clear costmap), etc

xiyuoh reviewed (commented):
Thanks for working on this!  Support for perform actions in free fleet is definitely valuable and anticipated. The various components of the action plugin are implemented well and testing with provided examples are working as intended.

Looking at `RobotActionContext` more closely, I think we can provide users with greater accessibility to the robot's API. The current free fleet action doesn't seem to support users to command their robots via Nav2, when it's the primary interface between RMF and the robot.

I suggest bringing `self.tf_handler`, `self.map_name`, `self.battery_soc` and `self.zenoh_session` into `RobotActionContext`, as well as import `free_fleet.ros2_types` into `action.py`. This would allow users to implement actions involving Nav2 commands (e.g. robot moves through a multi-point path, similar to demos cleaning task), and also have access to robot state updates.

aaronchongth commented:
That's a good point, however with the actions abstracted outside of any specific `RobotAdapter`, it becomes an issue to pass in different types of TF handlers (ROS 1 or ROS 2, or anything else that handles poses).

I think I'd prefer that all the necessary information sources (subscriptions, custom controls, etc) to be implemented into their action themselves. If there is a need for the positions, they can get it via a new `TfHandler` or setting up a new subscription. This will ensure that actions are standalone, and do not rely on the information provided by other sources, just in case they may be delayed or not handled at the frequency needed. This extends to the map name, battery soc, other pieces of information, etc.

For the example of running a Nav2 BT from an action, the action will need to define the proper Zenoh interfaces, as well as ensure that the zenoh bridge config is set up with the required ROS 2 navigation actions. I'll set up an example for that down the road too.

aaronchongth commented:
I've added a `get_pose` callable instead, which either returns the pose (as a 3 element list that RMF normally uses), or fails and returns a None. Added it's usage in the `hello_world` example as well

xiyuoh reviewed (approved):
Thanks for making the changes! Looks great with added support for users to retrieve fleet-relevant info. LGTM
